### ISSUE TITLE
NullReferenceException in ExternallyControlledLifetimeManager

### ISSUE THREAD

**Issue by appel1** [unitycontainer/abstractions#114]
```C#
[TestMethod]
public void NullReferenceExceptionWhenGettingAndSettingsValuesFromMultipleThreads()
{
var e = new ExternallyControlledLifetimeManager();

Parallel.For(
  0,
  100000,
  i =>
  {
    if (i % 2 == 0)
      e.SetValue(new Foo());
    else
    {
      var foo = e.GetValue();
    }
  });

public class Foo
{
  public int[] Value { get; } = new int[1024 * 1024];
}
}
```

If we resolve values registered with ExternallyControlledLifetimeManager from multiple threads we quite frequently get a NullReferenceException in ExternallyControlledLifetimeManager.GetValue. There are no locks as far as I can tell anywhere so it boils down to what happens with the code above. 

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Unity.Lifetime.ExternallyControlledLifetimeManager.GetValue(ILifetimeContainer container)
   at Unity.Strategies.LifetimeStrategy.PreBuildUp(BuilderContext& context)
   at Unity.UnityContainer.<>c.<.ctor>b__73_1(BuilderContext& context)
   at Unity.UnityContainer.Unity.IUnityContainer.Resolve(Type type, String name, ResolverOverride[] overrides)
   at Unity.UnityContainerExtensions.Resolve[T](IUnityContainer container, String name, ResolverOverride[] overrides)
```

This happens with Unity.Container 5.11.1.0 and Unity.Abstractions 4.2.0.0. 
It did not happen with Unity.Container 5.8.13.0 and Unity.Abstractions 3.3.1.0.

**Comment by ENikS** [comment]
The externally controlled lifetime manager is not supposed to be used like that. It is initialized only once during instance registration and holds on to that value for the rest of its life. 

Theoretically, if the object it holds reference to is already collected, it might create another one, but it all should happened inside the Unity.

**Comment by appel1** [comment]
This was just to show how the error occurs. The same thing happens if you resolve something registered using ExternallyControlledLifetimeManager from multiple threads as you can sort of see from the stack trace.

**Comment by ENikS** [comment]
If you could create a test case with the container I will investigate.

**Comment by appel1** [comment]
Attached a small program that triggers the issue. Start the release build without a debugger attached and press the start button. On my machine it usually takes somewhere between 2-10 iterations until you get a NullReferenceException.

[WpfApp2.zip](https://github.com/unitycontainer/abstractions/files/3901252/WpfApp2.zip)

**Comment by appel1** [comment]
A possible improvement would be to make ExternallyControlledLifetimeManager a SynchronizedLifetimeManager to avoid creating multiple instances unnecessarily.

**Comment by ENikS** [comment]
I agree, I’ll make the change in next version

**Comment by ENikS** [comment]
After playing with it a bit, I think there are more to it than meets the eye. I see two cases where `ExternallyControlledLifetimeManager` could be used. 
- external instance
- **type** or **factory** registration

In case of instance registration, container can use the instance until it is collected by the GC. Once collected, container should throw an exception if the instance is requested. This behavior should be enforced because container does not know how to create new instance (if at all possible) and not throwing the exception will destabilize environment and create invalid state.

If **type** or **factory** is registered with this manager it should only create new instance if it is either never been created or already collected. (This scenario is useful in all kinds of cache scenarions)

I do not think this behavior could be implemented in one manager. I think `ExternallyControlledLifetimeManager` should be reserved for instance registrations and new manager `WeakReferenceLifetimeManager` should be created to cover  **type** and **factory** registrations.

**Comment by ENikS** [comment]
I've created [this issue](https://github.com/unitycontainer/abstractions/issues/126) to track progress of the fix

**Comment by appel1** [comment]
Would still be useful to fix ExternallyControlledLifetimeManager in a 5.x release and then do your major refactoring in 6.x.

**Comment by ENikS** [comment]
I do not believe it is broken. It was never intended to be used as you do, look at the [original implementation](https://github.com/unitycontainer/unity/blob/3f9efb0919dc78aa21fe323dfdfb458848339491/source/Unity/Src/Lifetime/ExternallyControlledLifetimeManager.cs). 

As I suggested, create your own `WeakReferenceLifetimeManager` and perhaps send a PR to develop branch?

**Comment by appel1** [comment]
The original implementation is thread-safe and so works just fine even if you resolve from multiple threads in parallel.

**Comment by ENikS** [comment]
After thinking a bit more about it, I think you are right.

**Comment by ENikS** [comment]
Fixed in 5.11.3
